File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.js
/**
 * @param {string} s
 * @return {number}
 */

 //map 사용
// var firstUniqChar = function(s) {
    
//     let newMap = new Map();

//     //전부 다 돌고
//     for(let idx = 0 ; idx<s.length ; idx++){
//         if(newMap.has(s[idx])){
//             let value = newMap.get(s[idx])
//             newMap.set(s[idx],value + 1)
//         } else {
//             newMap.set(s[idx], 1);
//         }
        
//     }

//     for(let idx = 0;  idx<s.length ;idx++){
//         if(newMap.get(s[idx]) ==1){
//             return idx;
//         }
//     }

//     //반복되는 것이 없다면, -1
//     return -1
// };

var firstUniqChar = function(s){
    let mp = {}

    for(let char of s){
        mp[char] = !mp[char] ? 1 : mp[char] +1;
    }

    for(let i = 0 ; i < s.length ;i++){
        if(mp[s[i]] == 1) return i;
    }

    return -1
}

File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.go
func firstUniqChar(s string) int {
    // 맵 생성: 문자(rune)를 키로, 등장 횟수를 값으로 저장
    newMap := make(map[rune]int)

    // 첫 번째 반복문: 각 문자의 등장 횟수 계산
    for _, v := range s {
        newMap[v]++
    }

    // 두 번째 반복문: 처음 등장 횟수가 1인 문자의 인덱스 반환
    for i, v := range s {
        if newMap[v] == 1 {
            return i
        }
    }

    // 모든 문자가 반복되면 -1 반환
    return -1
}